src/error.rs

```diff
@@ -0,0 +1,186 @@
+            AppErrorKind::NotFound => write!(f, "{}", "Not Found"),
+            AppErrorKind::MethodNotAllowed => write!(f, "{}", "Method Not Allowed"),
+            AppErrorKind::UnprocessableEntity => write!(f, "{}", "Unprocessable Entity"),
+            AppErrorKind::TooManyRequests => write!(f, "{}", "Too Many Requests"),
+            AppErrorKind::InternalServerError => write!(f, "{}", "Internal Server Error"),
+            AppErrorKind::RocketError(ref error) => write!(f, "{:?}", error),
+
+            AppErrorKind::InvalidEmailParams(ref error) => {
+                if let Some(error) = error {
+                    write!(f, "{:?}", error)
+                } else {
+                    write!(f, "{}", "Error validating email params.")
+                }
+            }
+
+            AppErrorKind::InvalidProvider(ref provider) => write!(f, "{}", provider),
+            AppErrorKind::ProviderError {
+                ref name,
+                ref description,
+            } => write!(f, "{} error: {:?}", name, description),
+            AppErrorKind::EmailParsingError(ref error) => write!(f, "{:?}", error),
+
+            AppErrorKind::BounceError {
+                ref address,
+                ref bounce,
+                ref description,
+            } => {
+                if let Some(bounce) = bounce {
+                    write!(
+                        f,
+                        "address: {}, bounce: {:?}, description: {}",
+                        address, bounce, description
+                    )
+                } else {
+                    write!(f, "address: {}, description: {}", address, description)
+                }
+            }
+
+            AppErrorKind::DbError(ref error) => write!(f, "{:?}", error),
+
+            AppErrorKind::NotImplemented => write!(f, "{}", "Feature not implemented"),
+        }
+    }
+}
+
+impl AppErrorKind {
+    /// Return a rocket response Status to be rendered for an error
+    pub fn http_status(&self) -> Status {
+        match *self {
+            AppErrorKind::NotFound => Status::NotFound,
+            AppErrorKind::MethodNotAllowed => Status::MethodNotAllowed,
+            AppErrorKind::UnprocessableEntity => Status::UnprocessableEntity,
+            AppErrorKind::TooManyRequests => Status::TooManyRequests,
+            AppErrorKind::InternalServerError => Status::InternalServerError,
+            AppErrorKind::DbError(ref _0) => Status::ServiceUnavailable,
+            AppErrorKind::BadRequest | _ => Status::BadRequest,
+        }
+    }
+}
+
+impl From<AppErrorKind> for AppError {
+    fn from(kind: AppErrorKind) -> AppError {
+        Context::new(kind).into()
+    }
+}
+
+impl From<Context<AppErrorKind>> for AppError {
+    fn from(inner: Context<AppErrorKind>) -> AppError {
+        AppError { inner }
+    }
+}
+
+/// Generate HTTP error responses for AppErrors
+impl<'r> Responder<'r> for AppError {
+    fn respond_to(self, request: &Request) -> response::Result<'r> {
```

Comment:
I'm just thinking out loud here, so we don't need to worry for this PR, but seeing this suggests that maybe in time we might want to have different implementations of this for error payloads that contain extra fields (e.g. validation errors where ultimately we want to list the invalid params etc). Or maybe we can do all that in this one implementation by having some kind of generic extra-fields mechanism. Dunno.